front/pages/api/lookup.ts

```diff
@@ -0,0 +1,174 @@
+});
+
+const WorkspaceLookupSchema = t.type({
+  workspace: t.string,
+});
+
+const ResourceType = t.union([t.literal("user"), t.literal("workspace")]);
+
+async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse<WithAPIErrorResponse<LookupResponseBody>>
+): Promise<void> {
+  if (req.method !== "POST") {
+    return apiError(req, res, {
+      status_code: 405,
+      api_error: {
+        type: "method_not_supported_error",
+        message: "Only POST requests are supported",
+      },
+    });
+  }
+
+  const bearerTokenRes = await getBearerToken(req);
+  if (bearerTokenRes.isErr()) {
+    return apiError(req, res, {
+      status_code: 401,
+      api_error: {
+        type: "not_authenticated",
+        message: "The request does not have valid authentication credentials",
+      },
+    });
+  }
+
+  if (bearerTokenRes.value !== config.getRegionResolverSecret()) {
+    return apiError(req, res, {
+      status_code: 401,
+      api_error: {
+        type: "invalid_basic_authorization_error",
+        message: "Invalid token",
+      },
+    });
+  }
+  const { resource } = req.query;
+  const resourceValidation = ResourceType.decode(resource);
+  if (isLeft(resourceValidation)) {
+    return apiError(req, res, {
+      status_code: 400,
+      api_error: {
+        type: "invalid_request_error",
+        message: "Invalid resource type. Must be 'user' or 'workspace'",
+      },
+    });
+  }
+
+  let response: LookupResponseBody | null = null;
+  switch (resourceValidation.right) {
+    case "user":
+      {
+        const bodyValidation = UserLookupSchema.decode(req.body);
+        if (isLeft(bodyValidation)) {
+          const pathError = reporter.formatValidationErrors(
+            bodyValidation.left
+          );
+          return apiError(req, res, {
+            status_code: 400,
+            api_error: {
+              type: "invalid_request_error",
+              message: `Invalid request body for user lookup: ${pathError}`,
+            },
+          });
+        }
+        response = await handleLookupUser(bodyValidation.right);
+      }
+      break;
+    case "workspace": {
+      const bodyValidation = WorkspaceLookupSchema.decode(req.body);
+      if (isLeft(bodyValidation)) {
+        const pathError = reporter.formatValidationErrors(bodyValidation.left);
+        return apiError(req, res, {
+          status_code: 400,
+          api_error: {
+            type: "invalid_request_error",
+            message: `Invalid request body for user lookup ${pathError}`,
+          },
+        });
+      }
+      response = await handleLookupWorkspace(bodyValidation.right);
+    }
+  }
+  res.status(200).json(response);
+  return;
+}
+
+async function handleLookupUser(
+  userLookup: t.TypeOf<typeof UserLookupSchema>
+): Promise<UserLookupResponse> {
+  // Check if user exists
+  const user = await fetchUserWithAuth0Sub(userLookup.user.sub);
+
+  // Check for pending invitations
+  const pendingInvite =
```

Comment:
pendingInvite and workspaceWithVerifiedDomain can be pulled concurrentlyh with a Promise.all